File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10000000)

class Node:

    def __init__(self,key):
      self.key = key 
      self.left = self.right = None

class BST:

  def __init__(self):
    self.root = None 
    self.left = self.right = None

  def find_parent(self, key):
    # 들어갈 자리의 부모 노드 찾기 

    parent = None
    child = self.root

    while child is not None:
      if key < child.key:
        parent , child = child , child.left 
      else:
        parent , child = child , child.right
    return parent

  def insert(self, key):
    p = self.find_parent(key)
    if p == None:
      self.root = Node(key)
    else:
      if key < p.key:
        p.left = Node(key)
      else:
        p.right = Node(key)

tree = BST()

while True:
  inputs = input().rstrip()
  if len(inputs) == 0:
    break
  else:
    key = int(inputs)
  tree.insert(key)

def posterial(node):
  if node.left != None:
    posterial(node.left)
  if node.right != None:
    posterial(node.right)
  print(node.key)

posterial(tree.root)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.js
const fs = require('fs');
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
let inputs = fs.readFileSync(filePath, 'utf-8').trim().split('\n');

const preorder = inputs.map(Number); // [50,30,24,5,28,45,98,52,60]

/* BST = (M,L,R) */
function BST(key) {
  this.root = key;
  this.left = null;
  this.right = null;
}

function buildTree(preorder, start, end) {
  const rootIdx = start;
  const root = new BST(preorder[rootIdx]);

  if (start > end) {
    return null;
  }

  let splitIdx = start + 1;
  while (preorder[splitIdx] < preorder[rootIdx] && splitIdx <= end) {
    splitIdx++;
  }

  root.left = buildTree(preorder, rootIdx + 1, splitIdx - 1);
  root.right = buildTree(preorder, splitIdx, end);

  return root;
}

const START = 0;
const END = preorder.length - 1;
const tree = buildTree(preorder, START, END);

function postOrder(tree) {
  if (tree.left) {
    postOrder(tree.left);
  }
  if (tree.right) {
    postOrder(tree.right);
  }

  console.log(tree.root);
}

postOrder(tree);
